File: buildSrc/src/main/java/JarPostProcessorPlugin.java
import groovy.json.JsonOutput;
import groovy.json.JsonSlurper;

import io.github.pacifistmc.forgix.plugin.ForgixMergeExtension;
import io.github.pacifistmc.forgix.plugin.MergeJarsTask;

import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.objectweb.asm.*;
import org.objectweb.asm.tree.*;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;
import java.util.jar.JarOutputStream;
import java.util.zip.Deflater;

@SuppressWarnings("unused")
public class JarPostProcessorPlugin implements Plugin<Project> {
	@Override
	public void apply(Project project) {
		if(!project.getRootProject().equals(project)) {
			throw new IllegalStateException("apply to root project please");
		}

		boolean stripLVTs = Boolean.parseBoolean(project.findProperty("strip_lvts").toString());

		ForgixMergeExtension forgix = project.getExtensions().getByType(ForgixMergeExtension.class);

		project.getTasks().register("squishJar", jar -> {
			MergeJarsTask mergeJars = project.getTasks().named("mergeJars", MergeJarsTask.class).get();
			jar.dependsOn(mergeJars);
			mergeJars.finalizedBy(jar);
			jar.setGroup("forgix");

			jar.getInputs().files(forgix.getOutputDir(), forgix.getMergedJarName());
			File output = new File(forgix.getOutputDir(), forgix.getMergedJarName());

			jar.doLast(ignore -> {
				Map<String, byte[]> entries = readJar(output);
				output.delete();

				entries.replaceAll((name, bytes) -> {
					if(name.endsWith(".class") && stripLVTs) {
						return stripLVTsFrom(name, bytes);
					}
					if(name.endsWith(".json") || name.endsWith(".mcmeta")) {
						return minifyJson(bytes);
					}
					return bytes;
				});

				entries.entrySet().removeIf(entry -> entry.getKey().startsWith("META-INF/services/"));

				try(JarOutputStream out = new JarOutputStream(Files.newOutputStream(output.toPath()))) {
					out.setLevel(Deflater.BEST_COMPRESSION);
					entries.forEach((name, bytes) -> {
						try {
							out.putNextEntry(new JarEntry(name));
							out.write(bytes);
							out.closeEntry();
						} catch (IOException e) {
							throw new RuntimeException("Failed to write jar entry: " + name, e);
						}
					});
				} catch (IOException e) {
					throw new RuntimeException("Failed to write jar file: " + output, e);
				}
			});
		});
	}

	private static Map<String, byte[]> readJar(File file) {
		Map<String, byte[]> entries = new LinkedHashMap<>();
		try(JarFile jarFile = new JarFile(file)) {
			jarFile.entries().asIterator().forEachRemaining(entry -> {
				if(entry.isDirectory()) return;
				try {
					byte[] data = jarFile.getInputStream(entry).readAllBytes();
					entries.put(entry.getName(), data);
				} catch (IOException e) {
					throw new RuntimeException("Failed to read jar entry: " + entry.getName(), e);
				}
			});
		} catch(Throwable t) {
			throw new RuntimeException("Failed to read jar file: " + file, t);
		}
		return entries;
	}

	private static byte[] stripLVTsFrom(String name, byte[] classBytes) {
		if(name.startsWith("dev/rdh/createunlimited/shadow")) return classBytes;

		try {
			ClassNode classNode = new ClassNode();
			new ClassReader(classBytes).accept(classNode, 0);

			classNode.methods.forEach(methodNode -> {
				List<LocalVariableNode> lvt = methodNode.localVariables;
				if(lvt != null) {
					lvt.clear();
				}
				List<ParameterNode> parameters = methodNode.parameters;
				if(parameters != null) {
					parameters.clear();
				}
			});

			ClassWriter classWriter = new ClassWriter(0);
			classNode.accept(classWriter);
			return classWriter.toByteArray();
		} catch(Throwable t) {
			throw new RuntimeException("Failed to process class file: " + name, t);
		}
	}

	private static byte[] minifyJson(byte[] jsonBytes) {
		return JsonOutput.toJson(new JsonSlurper().parse(jsonBytes)).getBytes();
	}
}

File: buildSrc/src/main/kotlin/JarPostprocessorPlugin.kt
import groovy.json.JsonOutput
import groovy.json.JsonSlurper

import io.github.pacifistmc.forgix.plugin.ForgixMergeExtension
import io.github.pacifistmc.forgix.plugin.MergeJarsTask

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.ProjectConfigurationException

import org.objectweb.asm.ClassReader
import org.objectweb.asm.ClassWriter
import org.objectweb.asm.tree.ClassNode

import java.io.File
import java.util.jar.JarFile

import org.gradle.kotlin.dsl.*
import java.util.jar.JarEntry
import java.util.jar.JarOutputStream
import java.util.zip.Deflater

class JarPostprocessorPlugin : Plugin<Project> {

	override fun apply(project: Project) {
		project.afterEvaluate {
			if(rootProject != this) {
				throw ProjectConfigurationException("JarPostProcessorPlugin can only be applied to the root project", IllegalStateException())
			}

			val stripLVTs = findProperty("strip_lvts").toString().toBoolean()
			val forgix: ForgixMergeExtension = extensions.getByType<ForgixMergeExtension>()

			tasks.register("squishJar") {
				val mergeJarsTask = tasks.withType<MergeJarsTask>()["mergeJars"]
				dependsOn(mergeJarsTask)
				mergeJarsTask.finalizedBy(this)
				group = "forgix"
				val output = File(forgix.outputDir, forgix.mergedJarName)
				inputs.files(output)

				doLast {
					val entries = readJar(output)
					output.delete()
					entries.replaceAll { name, data ->
						if(name.endsWith(".class") && !name.startsWith("dev/rdh/createunlimited/shadow")) {
							val deAnnotationed = removeUnnecessaryAnnotationsFrom(data)
							if(stripLVTs) {
								stripLVTsFrom(deAnnotationed)
							} else {
								deAnnotationed
							}
						} else if(name.endsWith(".json")) {
							minifyJson(data)
						} else {
							data
						}
					}

					entries.keys.removeIf { it.startsWith("META-INF/services/") }

					JarOutputStream(output.outputStream()).use {
						it.setLevel(Deflater.BEST_COMPRESSION)
						entries.forEach { (name, data) ->
							it.putNextEntry(JarEntry(name))
							it.write(data)
							it.closeEntry()
						}
					}
				}
			}
		}
	}

	private fun readJar(file: File): MutableMap<String, ByteArray> {
		val entries: MutableMap<String, ByteArray> = LinkedHashMap()
		JarFile(file).use {
			it.entries().asIterator().forEach { entry ->
				if (entry.isDirectory()) return@forEach
				val data = it.getInputStream(entry).readAllBytes()
				entries[entry.name] = data
			}
		}
		return entries
	}

	private fun stripLVTsFrom(classBytes: ByteArray): ByteArray {
		val classNode = ClassNode()
		ClassReader(classBytes).accept(classNode, 0)

		classNode.methods.forEach {
			it.localVariables?.clear()
			it.parameters?.clear()
		}

		val classWriter = ClassWriter(0)
		classNode.accept(classWriter)
		return classWriter.toByteArray()
	}

	private fun removeUnnecessaryAnnotationsFrom(classBytes: ByteArray): ByteArray {
		val classNode = ClassNode()
		ClassReader(classBytes).accept(classNode, 0)

		classNode.methods?.forEach { method ->
			method.invisibleAnnotations?.clear()
		}

		val classWriter = ClassWriter(0)
		classNode.accept(classWriter)
		return classWriter.toByteArray()
	}

	private fun minifyJson(jsonBytes: ByteArray): ByteArray {
		return JsonOutput.toJson(JsonSlurper().parse(jsonBytes)).toByteArray()
	}
}